src/services/ContributingDeveloperAuditService/scm/GitHub/api/GitHubApiClient.ts

```diff
@@ -0,0 +1,215 @@
+      baseURL: baseUri,
+      headers: {
+        accept: "application/vnd.github+json",
+        Authorization: `Bearer ${githubPAT}`,
+      },
+    });
+
+    client.interceptors.request.use(
+      (request) => {
+        if (request.data) {
+          soosLogger.verboseDebug(
+            apiClientName,
+            `Request URL: ${request.method?.toLocaleUpperCase()} ${request.url}`,
+          );
+          if (request.params) {
+            soosLogger.verboseDebug(
+              apiClientName,
+              `Request Params: ${JSON.stringify(request.params)}`,
+            );
+          }
+          soosLogger.verboseDebug(apiClientName, `Request Body: ${JSON.stringify(request.data)}`);
+        }
+        return request;
+      },
+      (rejectedRequest) => {
+        return Promise.reject(rejectedRequest);
+      },
+    );
+
+    client.interceptors.response.use(
+      async (response) => {
+        soosLogger.verboseDebug(apiClientName, `Response Body: ${JSON.stringify(response.data)}`);
+        if (response.config.url?.includes("per_page")) {
+          return await GitHubApiClient.handlePagination(response, client);
+        }
+        return response;
+      },
+      async (error) => {
+        const { config, response } = error;
+        const maxRetries = 3;
+        config.retryCount = config.retryCount || 0;
+
+        if (response?.status === 429 && config.retryCount < maxRetries) {
+          config.retryCount += 1;
+          soosLogger.verboseDebug(
+            apiClientName,
+            `Retrying request (${config.retryCount}) after 3 minutes due to 429 status.`,
+          );
+          await sleep(180000);
+          return client(config);
+        }
+
+        if (response?.status) {
+          soosLogger.verboseDebug(apiClientName, `Response Status: ${response.status}`);
+        }
+        return Promise.reject(error);
+      },
+    );
+
+    return client;
+  }
+
+  private static async handlePagination(
+    response: AxiosResponse,
+    client: AxiosInstance,
+  ): Promise<AxiosResponse> {
+    let data = response.data;
+    let nextUrl = GitHubApiClient.getNextPageUrl(response);
+    while (nextUrl) {
+      soosLogger.verboseDebug("Handling pagination for ", response.config.url);
+      soosLogger.verboseDebug("Next url being fetched for pagination", nextUrl);
+      const nextPageResponse = await client.get(nextUrl);
+      data = data.concat(nextPageResponse.data);
+      nextUrl = GitHubApiClient.getNextPageUrl(nextPageResponse);
+    }
+
+    return { ...response, data };
+  }
+
+  private static getNextPageUrl(response: AxiosResponse): string | null {
+    const linkHeader = response.headers["link"] as string | undefined;
+    const nextLink = linkHeader?.split(",").find((s) => s.includes('rel="next"'));
+    return nextLink
+      ? new URL(nextLink.split(";")[0].trim().slice(1, -1), response.config.baseURL).toString()
+      : null;
+  }
+
+  async getGithubOrgs(): Promise<GitHubOrganization[]> {
```

Comment:
I might be missing something and probably overlooked something when I was thinking about how this should work. Are we running this for every org the user is associated with? If so, I think we should require an organization to be passed into the script as a required arg.

Reply:
If im not mistaken the user can generate the token with limited view of organizations and repositories (so they can choose that when they create it). But for sure we can add additional arguments to limit to certain organization names and even repositories.

Reply:
we chatted on slack and the end user will provide the organizationName on the command line